### DISCUSSION TITLE
Features from other internal tools apps

### DISCUSSION THREAD

**Discussion by JonL1** [directus/directus#10443] [Feature Requests]
Hi,

Moving here the conversation from discord.

As I said I'm really impressed with Directus. I was already impressed by the CMS features but now with granular permissions you can build nice internal tools also. However it is still lacking some features that make other non-CMS alternatives shine in the internal tools area.

I'm talking about Forestadmin, Jetadmin, Retool, Internal, Appsmith, Internal, etc

I'm creating a list here of features that would make great additions to the App. Not saying Directus has to mirror 1-to-1 everything that these tools have, but it would make it very interesting to use it also as an internal tool builder.

I might make assumptions that are wrong about directus so please bear with me and correct me if I'm wrong.

### Views

- These tools allow to build complex read/edit views with additional widgets. It's like a summary view of all the information related to a record and it's related data or a summary view of the data that is needed to complete a full workflow. 

![image](https://user-images.githubusercontent.com/1325712/145623664-c1bf2a9a-e32f-4289-b89b-6519a8a4b60f.png)

![image](https://user-images.githubusercontent.com/1325712/145622447-d9d43a7b-9bae-4a28-9f05-5a2b5806da89.png)

![image](https://user-images.githubusercontent.com/1325712/145623556-efc0e18e-dc7e-4376-90cc-e9551a19d4a4.png)

![image](https://user-images.githubusercontent.com/1325712/145625818-7cad0887-243e-4907-b270-548d4ce28db9.png)

- They also allow to hook external data(i.e. hubspot CRM). These could be achieved with virtual collections or something like that.

- They normally allow using JS bindings to show calculated fields and add certain logic when viewing data. 

![image](https://user-images.githubusercontent.com/1325712/145623995-2f2d5942-63eb-4ee2-8bc2-67edca764b4a.png)

![image](https://user-images.githubusercontent.com/1325712/145626167-7d29e526-538d-4cb2-a9de-df486c3d3476.png)

![image](https://user-images.githubusercontent.com/1325712/145626199-58eeedf5-c6b1-4c7a-ab55-1b7bda5994be.png)

- Inline editing. Editing fields directly from the table view. This would be just a nice addition. 

- Currently Directus seems to provide only an edit view when selecting a record. By default it should be a read only view and allow to edit using a button, keyboard shortcut or context menu.

### Workflow, actions, hooks, etc

These tools provide some way or another the option of processing a record besides CRUD operations. Some implement a no-code builder to create flows and others just a code editor and access to the needed data and variables.

 But they all provide in the UI adding buttons, icons, context menus, widgets, etc to invoke these functions. I believe directus provides webhooks but only for CUD operations. I can't select a record and generate an invoice for example.

![image](https://user-images.githubusercontent.com/1325712/145624463-3e0bfe4f-9626-4841-91d4-0130737d24d9.png)

![image](https://user-images.githubusercontent.com/1325712/145624336-d1d09e10-d9cc-4323-a099-26f449832d82.png)

![image](https://user-images.githubusercontent.com/1325712/145625901-b3c276ee-8afb-4c45-872f-ef18b7f23529.png)

![image](https://user-images.githubusercontent.com/1325712/145626488-6ca1586b-61f5-49be-be5d-90f6daf56724.png)

### Deployment

- They normally handle from the UI the migration of the config between environments( dev, staging, prod). I guess you could always dump the schema and restore it on another database but a streamlined way would be nice. Maybe via knex migrations.

**Comment by rijkvanzanten** [comment]
Great points all around!

Some thoughts at the top of my head:

### Views

> These tools allow to build complex read/edit views with additional widgets. It's like a summary view of all the information related to a record and it's related data or a summary view of the data that is needed to complete a full workflow. 

This is the Insights Module. That allows you to pull in graphs / relational data, and construct your own "views" (not to be confused with database views) or dashboards.

> They also allow to hook external data(i.e. hubspot CRM). These could be achieved with virtual collections or something like that.

This is definitely interesting. You could technically already "sync" data inwards by posting metrics into a custom collection, but having a dedicated "virtual collection" for reading data from external data providers would be very useful for this sort of use case.

> They normally allow using JS bindings to show calculated fields and add certain logic when viewing data. 

Using inline code is really powerful, but i'm wondering if there's any way we can solve for similar use cases without having to resort to "build it yourself" when it comes to this 🤔 

> Inline editing. Editing fields directly from the table view. This would be just a nice addition. 

💯 We've been meaning to add a Spreadsheet Layout (https://github.com/directus/directus/discussions/3874). Hopefully soon 🙂 

> Currently Directus seems to provide only an edit view when selecting a record. By default it should be a read only view and allow to edit using a button, keyboard shortcut or context menu.

I think this is a tad too opinionated. When it comes to a dashboard (like what we're doing on insights) then I'd agree. For data management, not so much 🤔

### Workflow, actions, hooks, etc

> These tools provide some way or another the option of processing a record besides CRUD operations. Some implement a no-code builder to create flows and others just a code editor and access to the needed data and variables.

We already have [a hooks system](https://docs.directus.io/extensions/hooks/) that allows you to do any custom logic on most API operations. I've been dreaming about having a way to configure those with commonly used actions (like "Send Email" or "Make API Request" or whatever) from a no-code editor. Having that effectively allows you to do a "IFTTT" style configuration for hooks.

> But they all provide in the UI adding buttons, icons, context menus, widgets, etc to invoke these functions. I believe directus provides webhooks but only for CUD operations. I can't select a record and generate an invoice for example.

Ohhh i like that idea. Configuring a hook that then gets a dedicated button to run manually would be a crazy powerful new thing to add.

### Deployment

> They normally handle from the UI the migration of the config between environments( dev, staging, prod). I guess you could always dump the schema and restore it on another database but a streamlined way would be nice. Maybe via knex migrations. 

Yeah! We already have the schema migration part figured out through the use of the `schema snapshot` and `schema apply` functions in the CLI. We wanna expose these same methods through the API as well, and later the app 🙂

**Reply by JonL1** [reply]
Thanks for the swift reply and great insights.

> This is the Insights Module. That allows you to pull in graphs / relational data, and construct your own "views" (not to be confused with database views) or dashboards.

Could you potentially build a dynamic insight panel and pass a record to it? Then we could build summary views or single record dashboards that bring in all the information needed for a specific record. Pair that with the hook buttons and virtual collections we talked about early and that makes the insights module even more powerful.

> Using inline code is really powerful, but i'm wondering if there's any way we can solve for similar use cases without having to resort to "build it yourself" when it comes to this 🤔

Maybe for simple visualisations like concatenating "fname" and "lname" into a virtual field, moustache notation or similar could be enough. The UI could provide autocomplete that reads from the schema the collections and fields. For more complex calculations a JS hook should be enough I guess. 

> I think this is a tad too opinionated. When it comes to a dashboard (like what we're doing on insights) then I'd agree. For data management, not so much 🤔

It is indeed and doesn't make much sense in a CMS UI context where you are creating, updating and deleting content and where most of the reading is done on the frontend. But for internal tools were you might want to give only read permissions to certain roles you would expect viewing first and then editing(if you are allowed).

But you are very right. If we could pass dynamically a record to an insight panel this could be solved easily. But still the UI/UX problem of what happens when I click on a row remains. By defect it goes to the edit form while I might want to create a read only role that can perform a search on a table and on click be sent to the single-record insight dashboard as they can't edit altogether.

>  Configuring a hook that then gets a dedicated button to run manually would be a crazy powerful new thing to add.

In these tools these buttons or links can be added in several places. In the table/card view as a virtual field that shows icon, link, button or as an icon(three dots) that show a dropdown and also in summary views or within certain widgets.

> We already have the schema migration part figured out through the use of the schema snapshot and schema apply functions in the CLI. We wanna expose these same methods through the API as well, and later the app 🙂

I'll check that out!

Thanks for time taken to review these ideas!

**Reply by rijkvanzanten** [reply]
> Could you potentially build a dynamic insight panel and pass a record to it? Then we could build summary views or single record dashboards that bring in all the information needed for a specific record. Pair that with the hook buttons and virtual collections we talked about early and that makes the insights module even more powerful.

I like where you're going with this!

We currently already have a "list" panel that allows you to pull in a subset of records that can be used for things like "Top 5 articles of 2021" sort of things.

I'm curious to learn more about the _Single record dashboard_. I'm imagining we could have a dropdown to select the particular item, and pass that on to all other panels to dynamically update the view 🤔 

> Maybe for simple visualisations like concatenating "fname" and "lname" into a virtual field, moustache notation or similar could be enough. The UI could provide autocomplete that reads from the schema the collections and fields. For more complex calculations a JS hook should be enough I guess.

Yeah! We do already have a "display template" system in place that's used all over the place (for things like page titles, table previews etc), so that could be used for the same purpose here. (It's also used in the "list" panel type now that i think of it!)

> By defect it goes to the edit form while I might want to create a read only role that can perform a search on a table and on click be sent to the single-record insight dashboard as they can't edit altogether.

True true! I've also been thinking through the possibility of adding Insights Panels to the Item Detail page 🤔 That way, Dashboards would be more global overviews, but you can still add little (readonly) graphs and widgets to the item detail page for this particular use case. That sounds like another great angle to solve for this particular use case.

---

Thanks for thinking through all this 🙂 Always super fun to imagine what the perfect UI/UX would be to make a crazy powerful setup like this as easy and user friendly as possible

**Reply by benhaynes** [reply]
> I'm curious to learn more about the _Single record dashboard_. I'm imagining we could have a dropdown to select the particular item, and pass that on to all other panels to dynamically update the view 🤔

Interesting idea! We're thinking about adding a global "date/range" filter to Insights so that you can update all time-series at once. I wonder if we could implement the ability to set other "dynamic variables" that can be controlled at the dashboard level and cascade throughout the Panels. 🤔  Then you could use that as a variable for an item ID... and then set that on the dashboard via a query string (so you can send people to the Dashboard, that is tailored based on the ID you referenced). I like this idea!

**Comment by JonL1** [comment]
> I'm curious to learn more about the Single record dashboard. I'm imagining we could have a dropdown to select the particular item, and pass that on to all other panels to dynamically update the view 🤔

As end-user of the Directus UI(not as an admin) I would expect them to probably spend 90% of the time browsing collection data. The single-record dashboard is probably something I would go to after searching or filtering a set of records so I would expect a direct access from collection layout view(table/card/calendar/etc)

An input on the dashboard would indeed tackle this if paired with typeahead search, but I guess it could be less used as people normally tend to work in collection views.

I'm just talking about single-record dashboards of course. For global ones I wouldn't expect anything else than finding them in the insights module with no access from elsewhere. 

>  I've also been thinking through the possibility of adding Insights Panels to the Item Detail page 🤔 That way, Dashboards would be more global overviews, but you can still add little (readonly) graphs and widgets to the item detail page for this particular use case. That sounds like another great angle to solve for this particular use case.

You can check the [live demo](https://app.forestadmin.com/livedemo) of forest admin. Clicking on any record on the table will take you to the single record view where you can see the following sub-views:

- Summary: Shows a record subset of data(and related), and the actions/hooks that are relevant to the process where that record is involved.
- Details: Shows all the raw data for the record.
- Explorer: Browse through all the m+1 related data.
- Analytics: Single-record insights.

Additionally something came to my mind. When I mentioned inline editing you forwarded me to a new feature for an editable grid. I was wondering why you decided on that approach instead of reusing the table view and just adding an inline edit input on double click or contextual menu. 

![image](https://user-images.githubusercontent.com/1325712/146058220-81f0648a-dcf8-481a-a1cd-d5388d55c9ad.png)

Is it because users want to tab cycle through records quickly to make changes? Users wouldn't be able to edit languages or other related data that way so unless it's a thing about power tab cycling or something I completely missed I wonder why people asked for it.

> Thanks for thinking through all this 🙂 Always super fun to imagine what the perfect UI/UX would be to make a crazy powerful setup like this as easy and user friendly as possible

It's a pleasure to be part of the fun. My only regret is that Vue is not my strength, but I work with a commercial framework that is heavily inspired by Vue so maybe one day I can help more that talk 😄

**Reply by rijkvanzanten** [reply]
> You can check the live demo of forest admin.

Oh my, that way of laying out the form / data points really fits in with the discussions @benhaynes and I have been having recently (we're envisioning what would happen if we'd use a same "infinite canvas" based setup that we have on dashboards for field configuration on the regular detail pages). Combined with a tabs-group and the ability to have panels on the detail page, and you could effectively layout a page that's just like that. 

> Is it because users want to tab cycle through records quickly to make changes?

Very good point. There's no reason we wouldn't be able to do a popover input as a context menu item in the "regular" tabular layout! I think there might be a handful of small UX benefits of a spreadsheet vs a table with rich displays like we have by default now, but that deserves some more R&D homework before I make any statements on that.

> It's a pleasure to be part of the fun. My only regret is that Vue is not my strength, but I work with a commercial framework that is heavily inspired by Vue so maybe one day I can help more that talk 😄

❤️

**Reply by benhaynes** [reply]
> I was wondering why you decided on that approach instead of reusing the table view and just adding an inline edit input on double click or contextual menu.

We could include this as a feature in the Table Layout (eg: edit mode, or right-click to edit a cell), but the Spreadsheet Layout was mostly different because it would be editing actual values... raw (not necessarily through an interface). This might not be as useful, but always down to discuss how these features overlap!

> It's a pleasure to be part of the fun. My only regret is that Vue is not my strength, but I work with a commercial framework that is heavily inspired by Vue so maybe one day I can help more that talk 😄

Would love this! Great ideas so far... we are only limited by our time! 😉

**Reply by benhaynes** [reply]
> Oh my, that way of laying out the form / data points really fits in with the discussions @benhaynes and I have been having recently (we're envisioning what would happen if we'd use a same "infinite canvas" based setup that we have on dashboards for field configuration on the regular detail pages). Combined with a tabs-group and the ability to have panels on the detail page, and you could effectively layout a page that's just like that.

Lot's to discuss here — lot's to potentially merge with the ideas behind Content and Insights (sharing Panels/Insights, grid layout, etc). We should really zoom out and think through how all of this comes together in an ideal world. 🤔

**Comment by JonL1** [comment]
> Oh my, that way of laying out the form / data points really fits in with the discussions @benhaynes and I have been having recently (we're envisioning what would happen if we'd use a same "infinite canvas" based setup that we have on dashboards for field configuration on the regular detail pages). Combined with a tabs-group and the ability to have panels on the detail page, and you could effectively layout a page that's just like that.

Glad you found it helpful. They have gone through a lot of iterations and feedback to get it right and I think they made a great job on the UI/UX for allowing people to build their dashboards and manage operations of their businesses. I guess you already saw the "layout editor" functionality with the drag&drop on the canvas.

> There's no reason we wouldn't be able to do a popover input as a context menu item in the "regular" tabular layout!

I guess this would kind of sort out the UX issue about defaulting to edit view on row click. If you allow inline editing on the regular tab layout you could explore safely the option of detail pages with infinite canvas without worrying about giving a default edit view to manage content as users could edit the most important data even before going into the detail view.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
